include/experimental/__p1673_bits/blas1_dot.hpp

```diff
@@ -59,21 +59,29 @@ struct is_custom_dot_avail : std::false_type {};
 template <class Exec, class v1_t, class v2_t, class Scalar>
 struct is_custom_dot_avail<
   Exec, v1_t, v2_t, Scalar,
-  std::enable_if_t<
-    std::is_same<
-      decltype(dot
-	       (std::declval<Exec>(),
-		std::declval<v1_t>(),
-		std::declval<v2_t>(),
-		std::declval<Scalar>()
-		)
-	       ),
-      Scalar
-      >::value
+  std::void_t<
```

Comment:
I'm not quite sure what you're trying to do here with `void_t`.  Are you trying to SFINAE out the body of this `struct` if `dot(Exec, v1_t, v2_1, Scalar)` is not well formed, or is this a redundant expression of `has_matching_return_type`?

Reply:
yes exactly, if this is not well formed it does not work. But if it, then I am also checking the return type is sam as `Scalar`. I was trying to address  your comment here: https://github.com/kokkos/stdBLAS/pull/101#discussion_r737626036

Reply:
Thanks @fnrizzi ! : - )  I think I see what the idea is here -- the `void_t` thing checks that the expression is well formed, and the `value` thing back-ports to C++ < 20 (else you would use `same_as` to constrain, then inherit from `true_type`).

One potential issue is that not all specializations of the struct inherit from either `true_type` or `false_type`.  However, I don't think that's a big deal for an implementation detail, as long as developers know to check `value` and not try to check inheritance.